### ISSUE TITLE
File automatically removed after uploading

### ISSUE THREAD

**Issue by SimonSimCity** [VeliovGroup/Meteor-Files#672]
### Give an expressive description of what is went wrong

A file, which is uploaded successfully can't be found when examining the destination folder. In one case I could even see it quickly appearing before it went away again.

### Version of `Meteor-Files` you're experiencing this issue

My `versions` file tells me that I'm using `ostrio:files@1.10.2`

### Version of `Meteor` you're experiencing this issue

Haven't checked any other - I'm currently on `METEOR@1.8`

### Where this issue appears? OS (Mac/Win/Linux)? Browser name and its version?

I'm effectively running on a mix of Mac and Linux for this example. I have a host system, which is Mac, holding the folder where the files should be stored. This folder is mounted into some virtual box machines, all running Ubuntu, which have a docker container holding the app. I'm using `meteor up` to deploy them. I sat this up to come closer to my staging environment, where I was seeing this issue before, which has multiple docker containers, running the app, on a single server, holding the folder.

### Is it *Client* or *Server* issue?

Looks like a server issue to me.

### Post *Client* and/or *Server* logs with enabled `debug` option, you can enable "debug" mode in [*Constructor*](https://github.com/VeliovGroup/Meteor-Files/wiki/Constructor)

```
[192.168.1.168]2019-02-05T12:03:17.702178628Z [FilesCollection] [File Start Method] Minimal.xlsx - wf9H2R63JSXNrhYGR
[192.168.1.168]2019-02-05T12:03:17.703847806Z [FilesCollection] [Upload] [DDP Start Method] Got #-1/1 chunks, dst: Minimal.xlsx
[192.168.1.168]2019-02-05T12:03:17.770289643Z [FilesCollection] [Upload] [DDP] Got #1/1 chunks, dst: Minimal.xlsx
[192.168.1.168]2019-02-05T12:03:17.819228020Z [FilesCollection] [Upload] [DDP] Got #-1/1 chunks, dst: Minimal.xlsx
[192.168.1.168]2019-02-05T12:03:17.852270100Z [FilesCollection] [Upload] [finish(ing)Upload] -> /data/imports/wf9H2R63JSXNrhYGR.xlsx
[192.168.1.168]2019-02-05T12:03:17.924674024Z [FilesCollection] [Upload] [finish(ed)Upload] -> /data/imports/wf9H2R63JSXNrhYGR.xlsx
[192.168.1.169]2019-02-05T12:03:17.600118615Z [FilesCollection] [_preCollectionCursor.observe] [changed]: wf9H2R63JSXNrhYGR
[192.168.1.169]2019-02-05T12:03:17.635168287Z [FilesCollection] [_preCollectionCursor.observe] [removed]: wf9H2R63JSXNrhYGR
[192.168.1.168]2019-02-05T12:03:18.036400123Z [FilesCollection] [_preCollectionCursor.observe] [removed]: wf9H2R63JSXNrhYGR
[192.168.1.168]2019-02-05T12:03:18.037418058Z [FilesCollection] [_preCollectionCursor.observe] [removeUnfinishedUpload]: wf9H2R63JSXNrhYGR
```

And I have to admit, it looks a lot similar to the problem, described in #324, but unfortunately I can't go back as far as to version 1.7.5 of this plugin.

Could also be related to #666 even though the author of this issue claims the file would exist and he doesn't show a trigger of `removeUnfinishedUpload`.

My problem is that I can't reproduce this on a regular basis, but only when running at least 3 of the instances - and even then it's only occasionally this happens. I guess I'm on a race condition here.

Occasionally it produces this log and leaves the file in place:

```
[192.168.1.168]2019-02-05T12:38:36.180408137Z [FilesCollection] [File Start Method] Minimal.xlsx - deAzpMh8QbhKpgeJe
[192.168.1.168]2019-02-05T12:38:36.182031228Z [FilesCollection] [Upload] [DDP Start Method] Got #-1/1 chunks, dst: Minimal.xlsx
[192.168.1.168]2019-02-05T12:38:36.229334437Z [FilesCollection] [Upload] [DDP] Got #1/1 chunks, dst: Minimal.xlsx
[192.168.1.168]2019-02-05T12:38:36.238273219Z [FilesCollection] [Upload] [DDP] Got #-1/1 chunks, dst: Minimal.xlsx
[192.168.1.168]2019-02-05T12:38:36.246799668Z [FilesCollection] [Upload] [finish(ing)Upload] -> /data/imports/deAzpMh8QbhKpgeJe.xlsx
[192.168.1.168]2019-02-05T12:38:36.286452749Z [FilesCollection] [Upload] [finish(ed)Upload] -> /data/imports/deAzpMh8QbhKpgeJe.xlsx
[192.168.1.169]2019-02-05T12:38:36.304769136Z [FilesCollection] [_preCollectionCursor.observe] [changed]: deAzpMh8QbhKpgeJe
[192.168.1.168]2019-02-05T12:38:36.314845185Z [FilesCollection] [_preCollectionCursor.observe] [changed]: deAzpMh8QbhKpgeJe
[192.168.1.169]2019-02-05T12:38:36.332151635Z [FilesCollection] [_preCollectionCursor.observe] [removed]: deAzpMh8QbhKpgeJe
[192.168.1.168]2019-02-05T12:38:36.346839057Z [FilesCollection] [_preCollectionCursor.observe] [removed]: deAzpMh8QbhKpgeJe
```

To nail it down, I've also tried to couple a version of the application, running on my mac in dev-mode, to the same database these 3 instances are running on, and I got it failing there as well - just much less times than in it's docker instances - all pointing to a race condition ...

**Comment by SimonSimCity** [comment]
Maybe worth noting that this happens specially when using [`redis-oplog`](https://github.com/cult-of-coders/redis-oplog), which replaces the MongoDB oplog.

**Comment by SimonSimCity** [comment]
Here's the result from the client ... seems to be relevant nonetheless

```
  core.js:108 [FilesCollection] [insert()]
  core.js:108 [FilesCollection] [FileUpload] [constructor]
  core.js:108 [FilesCollection] [insert] using WebWorkers
  core.js:108 [FilesCollection] [UploadInstance] [createStreams]
  upload.js:349 loadFile Minimal.xlsx: 154.962890625ms
  core.js:108 [FilesCollection] [UploadInstance] [sendEOF] false
  core.js:108 [FilesCollection] [link(qYRnHAFpbF33bWqDp, original)]
2 core.js:108 [FilesCollection] [link(qYRnHAFpbF33bWqDp, original)]
  core.js:108 [FilesCollection] [link(qYRnHAFpbF33bWqDp, original)]
2 core.js:108 [FilesCollection] [link(qYRnHAFpbF33bWqDp, original)]
  core.js:108 [FilesCollection] [UploadInstance] [end] Minimal.xlsx
  upload.js:301 insert Minimal.xlsx: 958.6259765625ms
2 core.js:108 [FilesCollection] [link(qYRnHAFpbF33bWqDp, original)]
7 core.js:108 [FilesCollection] [link(qYRnHAFpbF33bWqDp, original)]

**Comment by SimonSimCity** [comment]
Ok, I've digged into it and I think I'm now aware of what the problem is and how a solution could look like, which would not require a change on either the Meteor core or the redis-oplog extension. In both cases this remains a valid problem, in both cases it remains a problem of each of the implementations.

Let's take the case where we have two servers, where one of them (let's call him the first) is closer to the database than the other (let's call him the second).

Also take a look at the hook, described here https://github.com/VeliovGroup/Meteor-Files/blob/a5149baff36dde6d9a415f4ef261fc27402fe4af/server.js#L313-L336

In this example, the first will trigger the update. He will also be the one picking it up first, triggering the delete and running the delete hook.

Now finally the second one receives the information that the document has changed, but is (due to https://github.com/meteor/meteor/issues/10443 or https://github.com/cult-of-coders/redis-oplog/issues/311 - depending on if you use `redis-oplog` or not) unable to receive the document, because the first server already deleted it. He therefore skips the update hook - alright, that's fine. Now he receives the delete-hook, which he wants to execute. He will, neither here, be able to receive the document, which means, that the property `isFinished` will not exist, which in turn will result in that he starts removing the file.

My suggestion: Set `isFinished` to false and to a strict check for false, which will be negative in case the document already is deleted. In this case another server already should have taken the right choice at this condition.

**Comment by JonathanLehner** [comment]
doesn't work -_-

**Comment by JonathanLehner** [comment]
I mean i have the same issue. No way to fix it

**Comment by SimonSimCity** [comment]
@JonathanLehner Are you saying that neither applying the change #673 has helped you? If so, please create a minimal environment where you can reproduce the problem and I'll have a look into it.

**Comment by JonathanLehner** [comment]
It works now. Thanks a lot! It seems the "stream" setting for the upload caused problems.

**Comment by FrancoisYu** [comment]
@JonathanLehner May I know where to change the stream setting? I have the same problem.

**Comment by SimonSimCity** [comment]
@FrancoisYu If you want to use my approach just follow one of the approaches described in https://guide.meteor.com/writing-atmosphere-packages.html#overriding-atmosphere-packages. The first option is fairly easy - just clone my branch (https://github.com/SimonSimCity/Meteor-Files/tree/fix-auto-file-removal) into the `packages/` folder of your meteor installation.

@JonathanLehner if you resolved it using a different way, please let us know.

**Comment by JonathanLehner** [comment]
I changed the stream setting in
`let uploadInstance = Images.insert({
file: file,
transport: 'http', //faster
meta: {
//locator: self.props.fileLocator,
userId: 0
//needs to have accounts-base installed.. userId: Meteor.userId() //
Optional, used to check on server for file tampering
},
streams: 1,
chunkSize: 1000,
allowWebWorkers: true // If you see issues with uploads, change this to
false
}, false)
`
with that it worked

**Comment by SimonSimCity** [comment]
Rather a work-around than an actual fix of the problem because you have to add it on every command, right?

Have you also tried using the code I proposed?

**Comment by JonathanLehner** [comment]
I couldn't find the "packages" folder, so I haven't been able to add it. Do I need to create the folder?

**Comment by SimonSimCity** [comment]
If you don't already have gone through patching an atmosphere package locally you might have to create this folder ... yes ... Please see the docs I've referenced previously for further details.

**Comment by JonathanLehner** [comment]
Cannot replicate anymore. Now I got "Something went wrong! [sendEOF] method doesn't returned JSON! Looks like you're on Cordova app or behind proxy, switching to DDP transport is recommended." And nothing works anymore -_-

**Comment by SimonSimCity** [comment]
@JonathanLehner please check out the version before my change. Does this also happen there as well? If yes, does it also happen with the master-branch of this repository here?

**Comment by JonathanLehner** [comment]
Everything works now. Thanks a lot! One of my less experienced colleagues
apparently named another collection "Accounts" globally, which broke a lot
in the files package. It might be good to mention this somewhere, or to
check inside the package.

**Comment by jankapunkt** [comment]
@SimonSimCity while you describe this problem as kind of a race condition, where the "fatser one wins",  I experience the error for both servers, no matter if one is "closer" to the database. Can you confirm this for your setup, too?

**Comment by SimonSimCity** [comment]
@jankapunkt yes - very rarely though. It happens far more often (almost predictable) if one of the servers has a noticeably longer latency to the database.

**Comment by jankapunkt** [comment]
Fix merged via #673 feel free to reopen if the issue reappears or new issues have been introduced as a consequence of #673 that are related to this.
